stream/src/main/java/com/instaclustr/kafka/connect/stream/codec/CharDecoder.java

```
@@ -0,0 +1,220 @@
+        while (skipLeft > 0) {
+            long skipped = stream.skip(skipLeft);
+            skipLeft -= skipped;
+        }
+        streamOffset = numBytes;
+    }
+
+    @Override
+    public void close() throws IOException {
+        stream.close();
+    }
+
+    @Override
+    public List<Record<String>> next(int batchSize) throws IOException {
+        List<Record<String>> records = null;
+        int nread = 0;
+        // Reader over S3's stream appears to ready() == true even when EOF
+        // Local filesystem however is !ready() when EOF
+        // So we need to test both conditions here for EOF
+        while (reader.ready()) {
+            nread = reader.read(buffer, offset, buffer.length - offset);
+            log.debug("Read {} characters from stream", nread);
+            if (nread == -1) {
+                // End of stream
+                break;
+            } else if (nread == 0) {
+                // No character decoded from stream input
+                break;
+            }
+            // Must have nread > 0 due to constrained return value range of BufferedReader.read()
+            offset += nread;
+            if (records == null)
+                records = new ArrayList<>();
+
+            String line;
+            boolean foundOneLine = false;
+            do {
+                line = extractLine();
+                if (line != null) {
+                    foundOneLine = true;
+                    log.debug("Extracted a line from buffer");
+                    records.add(new Record<>(line, streamOffset, null, Schema.STRING_SCHEMA));
+
+                    if (records.size() >= batchSize) {
+                        log.debug("Return records in full batch: size=" + records.size());
+                        return records;
+                    }
+                }
+            } while (line != null);
+
+            if (!foundOneLine && offset == buffer.length) {
+                char[] newbuf = new char[buffer.length * 2];
```

If there was an extremely long line could this dynamic increase of the buffer length cause memory issues?

---

Great question!  Current behavior is a runtime OutOfMemoryError.  
Theoretically we can upperbound it by `Integer.MAX_SIZE`, and throw a nicer(?) error.  But it's still the same behavior.  
In practice, there is actually deeper problem: runtime JVM's heap size constraints how large an array can be, often less than `Integer.MAX_SIZE`, and it's unknowable: https://bugs.openjdk.org/browse/JDK-8246725

An alternative is to catch OutOfMemory and skip the superlong record.  At the moment I opt for fail fast.  If there's a need to "skip overflowing record", then it might be option to add.